pystiche/papers/ulyanov_et_al_2016/loss.py

```diff
@@ -0,0 +1,182 @@
+        elif impl_mode == "paper":
+            score_weight = 1e0  # FIXME paper only alpha
+        elif impl_mode == "master":
+            score_weight = 1e0
+        else:
+            raise NotImplementedError
+
+    if multi_layer_encoder is None:
+        multi_layer_encoder = ulyanov_et_al_2016_multi_layer_encoder()
+    encoder = multi_layer_encoder[layer]
+
+    return MSEEncodingOperator(encoder, score_weight=score_weight)
+
+
+class UlyanovEtAl2016GramOperator(GramOperator):
+    def __init__(
+        self, encoder: Encoder, impl_params: bool = True, **gram_op_kwargs: Any,
+    ):
+        super().__init__(encoder, **gram_op_kwargs)
+        self.normalize_by_num_channels = impl_params
+
+    def enc_to_repr(self, enc: torch.Tensor) -> torch.Tensor:
+        gram_matrix = super().enc_to_repr(enc)
+        if not self.normalize_by_num_channels:
+            return gram_matrix
+
+        num_channels = gram_matrix.size()[-1]
+        return gram_matrix / num_channels
+
+
+def ulyanov_et_al_2016_style_loss(
+    impl_mode: str = "paper",
+    mode: str = "texture",
+    multi_layer_encoder: Optional[MultiLayerEncoder] = None,
+    layers: Optional[Sequence[str]] = None,
+    layer_weights: Union[str, Sequence[float]] = "sum",
+    score_weight: float = None,
+    **gram_op_kwargs,
+):
+    if score_weight is None:
+        if impl_mode == "texturenetsv1":
+            score_weight = 1e3 if mode == "style" else 1e0
+        elif impl_mode == "paper":
+            score_weight = 1e0
+        elif impl_mode == "master":
+            score_weight = 1e0
+        else:
+            raise NotImplementedError
+
+    if multi_layer_encoder is None:
+        multi_layer_encoder = ulyanov_et_al_2016_multi_layer_encoder()
+
+    if layers is None:
+        if impl_mode == "texturenetsv1":
+            layers = ("relu_1_1", "relu_2_1", "relu_3_1", "relu_4_1", "relu_5_1")
+        elif impl_mode == "paper":
+            layers = ("relu_1_1", "relu_2_1", "relu_3_1", "relu_4_1", "relu_5_1")
+        elif impl_mode == "master":
+            layers = ("relu_1_1", "relu_2_1", "relu_3_1", "relu_4_1")
+        else:
+            raise NotImplementedError
+
+    def get_encoding_op(encoder, layer_weight):
+        return UlyanovEtAl2016GramOperator(
+            encoder, score_weight=layer_weight, **gram_op_kwargs
+        )
+
+    return MultiLayerEncodingOperator(
+        multi_layer_encoder,
+        layers,
+        get_encoding_op,
+        layer_weights=layer_weights,
+        score_weight=score_weight,
+    )
+
+
+def ulyanov_et_al_2016_regularization(
+    impl_mode: str = "paper",
+    score_weight: float = None,
+    **total_variation_op_kwargs: Any,
+):
+    if score_weight is None:
+        if impl_mode == "texturenetsv1":
+            return None
+        elif impl_mode == "paper":
+            return None
+        elif impl_mode == "master":
+            score_weight = 0  # FIXME: tvLoss only in master but weight 0
```

Comment:
Does that mean that the TV loss is not used at all? If that is the case please remove it completely.